### ISSUE TITLE
Suggestion: Add build server hook to mark commits / PRs as good or bad

### ISSUE THREAD

**Issue by vors** [pester/Pester#251]
appveyor provides free cloud based continuous integration for open sources projects.
Adding it will automatically build all PR and run tests. Which provide higher confidence in source quility.

Here is how it may look like https://ci.appveyor.com/project/vors/zlocation
Status will be reported back to readme file with badge, like here
https://github.com/vors/ZLocation

I can take care about setting it up (it's pretty easy).
My only concern is: which account I should use to do it.

Is it fine if I will use my personal account? 
(The only downside that I see so far is somewhere in README.md file will be link to https://ci.appveyor.com/project/vors/pester, not https://ci.appveyor.com/project/pester/pester)

**Comment by dlwyatt** [comment]
For Pester, you mean?  We already have a CI job set up.  @nohwnd has all the details.  :)

**Comment by vors** [comment]
@nohwnd does it build all PRs?

**Comment by nohwnd** [comment]
@vors Pester is built on Team City server provided by CodeBetter. It is packaged automatically in a nupkg and pushed to choco, nuget and powershell gallery. I am not sure if there is more things we need. At the moment I think we use it mainly to make the deployment of new versions easier. 

I don't think it builds PRs but I don't see the point either. We (or me at least) merge the versions locally and run the test suite every time.

As for the icon. I think it would be useless now, because all the builds that don't have correct version tags fail the build at the moment. But if I configure that correctly I can add icon to github. It will be green 99,99% percent of the time though. Pester is not "broken" often.

So to sum up, thanks for the offer, but from my point of view moving to another CI does not make sense at the moment.

**Comment by vors** [comment]
Awesome! I didn't know you have such a nice solution already. (I :heart: teamcity )

I see a lot of value in building PRs and use github hooks to mark commits as valid or failed.
It will encourage contribution for new people (like me) by increasing confidence in PRs correctness.
Also, we are all like automation, right? :)

![image](https://cloud.githubusercontent.com/assets/816680/5762917/a7f63802-9c9a-11e4-9e06-e5459fc8905e.png)

Here is an article by JetBrains about seting up build hooks for teamcity http://blog.jetbrains.com/teamcity/2013/02/automatically-building-pull-requests-from-github-with-teamcity/

**Comment by nohwnd** [comment]
@vors If you'd like to help with the TC setup that would be great. Three are few things I don't know how to setup. Like conditional publish on tagged commits only (that is tagged with git tag not TC tags).

**Comment by dlwyatt** [comment]
There's one thing that would be cool (though I don't know if CodeBetter or Appveyor offer us this capability), and that is automatically running the Pester tests on PowerShell 2.0, 3.0, 4.0 and 5.0.  (2.0 is a freebie; can be run anywhere that has the feature installed by using `powershell.exe -Version 2.0`, but the other three versions require multiple build agents.)

Right now I do this manually using VMs, but it's cumbersome and annoying, and frequently I get lazy and only test on 2.0 + X, where X happens to be whatever is on my development machine (currently 5.0).

**Comment by vors** [comment]
@nohwnd  Sure, I can help with it.
@dlwyatt Why it require multiply build agents? We can use one build agent with WMF 5 installed and run all PS version on it.

**Comment by dlwyatt** [comment]
How do you accomplish that?  Try running "powershell -version 3.0" or "powershell -version 4.0" on a system with WMF 5 installed.  It launches just fine, but when you look at `$psversiontable`, it'll just be running PowerShell 5.0.  I'm not aware of a way for PowerShell 3-5 to coexist on the same OS instance (though I have hopes that the container features in Windows 10 will make this possible.)

**Comment by nohwnd** [comment]
That's a great point Dave! Having this, the other points, like building PRs, would make sense as well. Unfortunately I don't see all the versions needed on TC. There is just 2k3 and 2k8 agent.

If the other solution doesn't have them either I can try to build it on Azure. Hope my credit there is sufficient to cover a month of building.

**Comment by dlwyatt** [comment]
I had briefly considered having Powershell.org host a TeamCity infrastructure for the PowerShell community.  However, when I started to look closer, I was having a hard time figuring out how to make that secure, and scrapped the idea.  (Anyone could commit malicious code to one of the source control repos, and the CI job would execute it, potentially having bad things happen to the CI server, steal other clients' API keys and whatnot, etc.)

Having our own setup is a different story, though.  Without the multi-tenant concerns (and with us around to make sure that no dodgy code gets merged into the Pester repo), we can just make it work.  We'd probably be looking at three or four cloud instances to pull this off:

- Windows 2012 R2 server with WMF 5.0.  (TeamCity server and some number of build agents.)
- Windows 2012 R2 server with WMF 4.0   (Build agents)
- Windows 2012 server with WMF 3.0 (Build agents)
- Optionally - Windows 2008 R2 server with WMF 2.0 (Build agents).  This would be the best way to do powershell 2.0 testing, but we could probably get away with just using "powershell.exe -Version 2.0" to save on costs.

Cost would depend on how many agents we want to run on each server, and what tier we'd used.  We're probably not supposed to use our MSDN credits for this sort of "production" use, but we could ask the Microsoft folks if they'd be willing to donate a few instances for this project.  With the way things are going lately, it's in their interest to have Pester working as well as possible anyway.  :)

**Comment by vors** [comment]
@dlwyatt you right, `powershell -Version 3.0` doesn't switch version. I didn't know.

I think that automated build verification for v2 and v4/v5 could be a good progress already.

Regards infrastructure for other projects: appveyor can be used as an infrustracture for all open source windows projects (includes PS community projects). It's pretty awesome product. Assuming we drop requirement about test runs an all versions.

**Comment by mwrock** [comment]
Fwiw I think this is a fantastic idea. Having been working with ruby for the past several months, it is very common for GitHub projects to have a hosted Travis service(the spiritual equivalent of appveyor on Linux) that runs prs through the test and you can see which are green. It's very helpful both for contributors and project owners.

It also can run everything through different versions of ruby and Linux distros and if you could get appveyor to run different PS versions that would be equally awesome.

**Comment by nohwnd** [comment]
@dlwyatt I will set that up as research just to see how that works.

**Comment by rismoney** [comment]
:+1: this idea and agree with @mwrock.  appveyor is the travis for windows at this point and I have never seen a failed appveyor experience yet (witnessed/used quite a bit now)

I firmly believe this will be a reflection point. Here are some warning signs (not to point fingers) but I see it with colleagues at work all the time:

> We (or me at least) merge the versions locally and run the test suite every time.
> Pester is not "broken" often.

The point is that the ability to add tests and people to get immediate feedback on tests without doing anything is a level of awesomeness.  the velocity and simplicity will improve quality.

Here is a pretty awesome thing we do against all commits in my private orgs repos- we ensure no tabs in files (2 space indents), trailing whitespace, ensure CRLF for windows and LF line endings for linux.  Boom totally repo cleanliness that noone ever thinks about.

Most importantly if it transfers preliminary code quality responsibilities back to the contributor.  The maintainer should be concerned with macro, deeper code diving, and other project responsibilities.

**Comment by vors** [comment]
Any updates?

**Comment by dlwyatt** [comment]
I think we can do this with CodeBetter.  They have the "report status change to GitHub" plugin installed, and we can set up the project to do whatever we like.  Unfortunately, they only have PowerShell v3 build agents at the moment, and I don't know if there's anything we can do about that.

If we get the opportunity to host our own TeamCity server and agents, we can improve on that situation.

**Comment by vors** [comment]
I can setup appveyor, it has PS v4. More impotent build config is just a `appveyor.yml` file in the root, so everybody will see up-to date build/test steps at every moment and can change them in PR, in case it's needed.

**Comment by vors** [comment]
I sent #257 
The question for me was: should I use pester version from choco or current version from repo as a unit test framework?
Both options are suspicious:
1) If version in choco has issue or don't support some feature, then correct tests would not pass. 
2) But it's possible to break current version completely and report green status for PR for instance.

I choose 2-nd option, because it's allow new feature development.

Now to enable integration on this repo you can:
1) register on appveyor.com with your github account
2) Add pester new project from https://github.com/pester/Pester (no setup needed, just create project and close the tab: all configuration happens in `.yml` file )
3) Accept PR #257

**Comment by dlwyatt** [comment]
I'm potentially open to switching CI platforms (though it's really @nohwnd who makes that call), but I don't want to rush into it.  I'm just getting the hang of TeamCity :) .

**Comment by vors** [comment]
don't worry, I didn't spent any significant time on that. Great that you figured out how to make it work on teamcity! I will close #257

**Comment by dlwyatt** [comment]
This should be all set.  I've done some more tinkering with TeamCity this evening.  PRs (and the master branch) will be tested on both PowerShell v2 and v3 (in parallel, which is neat), and the PRs will be marked as success or failure.

If CodeBetter sets up a build agent with WMF 4 and/or 5 at some point (or if we move to our own TeamCity servers), we can add tests for those versions as well.
